feat: new chart API

markov00 opened the pull request:
## Summary

### Todos

- [x] reenable legend values
- [x] reenable external cursor events/share cursor position
- [x] ~check  deselectedDataSeries: DataSeriesColorsValues[];~
- [x] `isCursorOnChart`
- [x] fix brush end

## Breaking Changes

- `SpecId`,`AxisId`, `AnnotationId` types are down-casted to a `string` type. The `getSpecId`, `getAxisId`m `getAnnotationId` methods still exist and but return just the same passed string until deprecated in a future version.
- The spec ids, previously `id`, `axisId`,`annotationId` etc are now aligned to use the same prop name: `id`
- The chart rendering status `data-ech-render-complete` and `data-ech-render-count` is no more at the root level of the `echChart` div, but on its child element: `echChartStatus` 

This PR refactor the current MobX implementation of elastic-chart in favour of redux and a better API to include new chart types.


### GlobalChartState
The new redux state is configured as the following:
```ts
export interface GlobalChartState {
  // an unique ID for each chart used by re-reselect to memoize selector per chart
  chartId: string;
  // true when all all the specs are parsed ad stored into the specs object
  specsInitialized: boolean;
  // true if the chart is rendered on dom
  chartRendered: boolean;
  // incremental count of the chart rendering
  chartRenderedCount: number;
  // the map of parsed specs
  specs: SpecList;
  // the chart type depending on the used specs
  chartType: ChartType | null;
  // a chart-type-dependant class that is used to render and share chart-type dependant functions
  internalChartState: InternalChartState | null;
  // the dimensions of the parent container, including the legend
  parentDimensions: Dimensions;
  // the state of the interactions
  interactions: InteractionsState;
  // external event state
  externalEvents: ExternalEventsState;
}
```

That global state maintain mainly the `specs` map (parsed from the `<SpecParser />` component, similarly to the previous MobX implementation. Each spec is now stored into a plain JS object where each key is the id of the `Spec`.
A `Spec` has now few more properties
```ts
export interface Spec {
  /** unique Spec identifier */
  id: string;
  /** Chart type define the type of chart that use this spec */
  chartType: ChartType;
  /** The type of spec, can be series, axis, annotation, settings etc*/
  specType: string;
}
```


### New Chart types

Each `ChartType` needs one or more `Spec` and needs to register its own `InternalChartState` class into the `initInternalChartState` function of `src/state/chart_state.ts`.
The `InternalChartState` class follow this interface:

```ts
export interface InternalChartState {
  // the chart type
  chartType: ChartType;
  // returns a JSX element with the chart rendered (lenged excluded)
  chartRenderer(containerRef: BackwardRef, forwardStageRef: RefObject<Stage>): JSX.Element | null;
  // true if the brush is available for this chart type
  isBrushAvailable(globalState: GlobalChartState): boolean;
  // true if the chart is empty (no data displayed)
  isChartEmpty(globalState: GlobalChartState): boolean;
  // return the list of legend items
  getLegendItems(globalState: GlobalChartState): Map<string, LegendItem>;
  // return the list of values for each legend item
  getLegendItemsValues(globalState: GlobalChartState): Map<string, TooltipLegendValue>;
  // return the CSS pointer cursor depending on the internal chart state
  getPointerCursor(globalState: GlobalChartState): string;
}
```




close https://github.com/elastic/elastic-charts/issues/75

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios
- [x] Each commit follows the [convention](https://github.com/elastic/elastic-charts/blob/master/CONTRIBUTING.md)

monfera commented:
First of all, great changes overall toward data flow simplification! 

There are a couple of places where some more uncoupling could be done, here's the main one: likely because of what chart types have been covered so far, `Chart` (`src/components/chart.tsx`) now depends on and builds an `xy_chart` (ie. stuff under `src/chart_types/xy_chart`) even when instantiating something like
```
<Chart>...
  <Pie...>
</Chart>
```
where `Pie` is in a separate directory under `/chart_types/` (even appears to work as long as structural typing is observed but won't use the code under `/chart_types/circline`)
  
 As `Chart` is a quite general name, and likely meant to be a top-level container for all things in `elastic-charts` from the API viewpoint, this coupling would ideally give way to a plurality of projection types incl. non-Cartesian ones. An alternative is to rename it eg. `CartesianChart`, and there may be other ways already thought out, or we all could brainstorm

monfera commented:
The other uncoupling question/suggestion is about whether certain computations could be separated from views that they are often associated with. For example, `get_cursor_band.ts` or `get_tooltip_values_highlighted_geoms.ts` seems to depend on axes, which are basically, things to render only. So maybe it could be split to two parts: the computational part would have the responsibility of returning which value, band, point etc. the cursor is over; and things like the Axis, Crosshair and Tooltip would merely be responsible for rendering. Ie. no sideways communication. I see the attraction, and how a coupling can evolve historically, eg. we need a crosshair and need to compute where the cursor is in data space, and then later we add tooltips, which can reuse some values already computed for the crosshair. It'd be neat to separate because eg. some charts may need tooltips that don't need crosshair.

Another example is `get_tooltip_values_highlighted_geoms.ts` and `tooltip.ts`. Here it's coupled, and data flow is happening sideways, because the tooltip takes formatting spec defaults from the axes (maybe other things happen too or there's a deeper explanation). Perhaps the default formatters could belong to the projection (each scale, eg. `xScale` and `yScale`) rather than the axes, or at least it'd be neat to reconsider data transfers between sibling components like tooltip vs axis. 

I'm not necessarily suggesting that instead of an `<Axis...>` tag, that currently specifies a `tickFormatter`, we mandate library users to use some newfangled, and for many users, hard to grasp `<Projection...>` tag under `<Chart>`, even though it'd make our technical work easier. But maybe it's possible to make it so that the `<Tooltip>` component need not depend on `<Axis>` because from the tooltip viewpoint, the `tickFormat` comes from "upward". For a hypothetical example, if later on, someone decides that the default `tickFormat` needs to come from somewhere else, eg. `<Projection>` or `<Formatters>` or `<Config>` or whatever, the tooltip code shouldn't need to change as where the spec or default comes _from_ should be outside the concern of the `Tooltip` component itself.  Btw. I might be off base as I'm new to the code, or there may be obstacles to such separation; or maybe such changes have already been planned, as it's a WIP PR

monfera commented:
Likely a known issue: apparently randomly, storybook examples fail:
```
Cannot read property 'legendPosition' of undefined
TypeError: Cannot read property 'legendPosition' of undefined
    at http://localhost:9001/main.08acef345097e16e8e69.bundle.js:10032:57
```
then it may work on another try. Could be some race condition

markov00 commented:
> There are a couple of places where some more uncoupling could be done, here's the main one: likely because of what chart types have been covered so far, `Chart` (`src/components/chart.tsx`) now depends on and builds an `xy_chart` (ie. stuff under `src/chart_types/xy_chart`) even when instantiating something like

You are right. as this is a WIP PR there still some refactoring that I need to go through before making it really available for review. Right not `Chart` depends on `xy_chart` for some event actions that needs to be moved into the XY chart process itself:
```
 this.chartStore.subscribe(() => {
      const state = this.chartStore.getState();
      const settings = getSettingsSpecSelector(state);
      if (this.state.legendPosition !== settings.legendPosition) {
        this.setState({
          legendPosition: settings.legendPosition,
        });
      }
      if (!isInitialized(state) || state.chartType !== ChartTypes.XYAxis) {
        return;
      }
      onElementOverCaller(state);
      onElementOutCaller(state);
      onElementClickCaller(state);
    });
```
IMHO I prefer to keep the main root component as `Chart` because I don't want to provide N multiple container classes that need to be changed depending on the chart type. When I will move the dependency into the XY chart there will no need to rename it in any other way.

monfera commented:
Definitely not news to you but I was curious, all MobX dependencies can now be removed  🎉
```
index b97abbc..5eabf14 100644
--- a/package.json
+++ b/package.json
@@ -139,8 +139,6 @@
     "fp-ts": "^1.14.2",
     "konva": "^2.6.0",
     "luxon": "^1.11.3",
-    "mobx": "^4.9.2",
-    "mobx-react": "^5.4.3",
     "newtype-ts": "^0.2.4",
```

monfera commented:
Hover, click events on legends don't show up in Storybook not sure if it's to do with the code change or something that changed in relation to Storybook (I suspect the latter due to test cases passing)
![image](https://user-images.githubusercontent.com/1548516/69259407-7f8f2200-0bbe-11ea-8d6f-6038895508a6.png)

monfera commented:
Brush select also doesn't report events in Storyboard

monfera commented:
Crosshairs events don't get reported in either [here](https://elastic.github.io/elastic-charts/?path=/story/interactions--crosshair-with-time-axis) or in this branch, so maybe it's not meant to or regressed earlier

monfera commented:
Render change action is unreported as well in Storybook

markov00 commented on yarn.lock:
```diff
@@ -2,52 +2,25 @@
 # yarn lockfile v1
```
it's committed, maybe I didn't understood your statement

markov00 replied:
ok I've seen what you mean

markov00 replied:
updated on cf1a93b

markov00 commented:
> Hover, click events on legends don't show up in Storybook not sure if it's to do with the code change or something that changed in relation to Storybook (I suspect the latter due to test cases passing)
> ![image](https://user-images.githubusercontent.com/1548516/69259407-7f8f2200-0bbe-11ea-8d6f-6038895508a6.png)

@monfera There was few missing things on this. I've fixed it on 50c6997
I've re-enabled the listeners + add the click event (not available in the current mobx version) and added few tests

monfera commented on .playground/webpack.config.js:
```diff
@@ -13,6 +13,7 @@ module.exports = {
         exclude: /node_modules/,
         options: {
           configFile: 'tsconfig.json',
+          transpileOnly: true,
```
This makes an amazing speed difference to Playground, great find! 🎉

markov00 replied:
👍

monfera commented on package.json:
```diff
@@ -153,10 +155,14 @@
     "mobx-react": "^5.4.3",
```
Please `yarn remove mobx mobx-react` 😄

markov00 replied:
done :D

monfera commented on src/chart_types/xy_chart/state/selectors/compute_brush_extent.ts:
```diff
@@ -0,0 +1,13 @@
+    // console.log('--- 8 computeBrushExtentSelector ---');
+    return computeBrushExtent(chartDimensions.chartDimensions, settingsSpecs.rotation, chartTransform);
+  },
+)((state) => state.chartId);
```
For consideration for a future, non-time-critical PR: `reselect` is great and one of its benefits is that it uses a lifting pattern (it lifts an ordinary function that works on regular values to a level where it'll work with the input of other selectors), though its syntax hides it. `reselect` has various alternatives so it'd be useful to eventually abstract it out and not necessarily use its syntax. The constituents could be broken up a bit more too. An example (not sure if I got all parts right):

```
const chartId = (state) => state.chartId;

const brushExtentSelector = select(
  (chartDimensions, chartTransform, settingsSpecs): BrushExtent =>
    computeBrushExtent(chartDimensions.chartDimensions, settingsSpecs.rotation, chartTransform),
)(computeChartDimensionsSelector, computeChartTransformSelector, getSettingsSpecSelector);

export const computeBrushExtentSelector = brushExtentSelector(chartId);
```

or with less verbose names:

```
const extent = select(
  (dimensions, transform, settings): BrushExtent =>
    brushExtent(dimensions.chartDimensions, settings.rotation, transform),
)(dimensions, transform, settings);

export const brushExtent = extent(chartId);
```
```

monfera replied:
... my phrasing was unclear, names that tell what the function is about, preferably with no or very few, widely used abbreviations, are a good thing and this can make function names a bit long. I think it's an OK price to pay as the code is more self-documenting, a little bit closer to literate programming, and it makes the code more accessible. It's the little pre-and postfix words like compute, get, selector etc. that can be eliminated (my example is guilty of some 😄)

Also, in my last example, the `getExtent` is not only noisily named; it also looks like a shallow wrapper over the Real Deal, which is `brushExtent()`. Sometimes such shallow (un)wrappers are useful, but too many of them may just add noise. So it's possible that a better example from me would've been a very slight refactor of `brushExtent` (=`computeBrushExtent`) such that we can do
```
const chartId = (state) => state.chartId;

const extent = select(brushExtent)(dimensions, transform, settings);

export const brushExtent = extent(chartId);
```
or, if the rotation really needs to be unwrapped,
```
const chartId = (state) => state.chartId;

const rotation = (transform) => transform.rotation;
const chartDimensions = (dimensions) => dimensions.chartDimensions;

const extent = select(brushExtent)(
  select(chartDimensions)(dimensions),
  select(rotation)(transform), 
  settings
);

export const brushExtent = extent(chartId);
```
Or some variation, eg. with `const rotation = select(transform => transform.rotation);`

Also, I forgot to add that other tools that do very similar lifting are Flyd, most.js, crosslink etc. so the data piping logic can be tweaked (eg. performance optimized) with no change to the userland code.

monfera replied:
Via discussion, resolved as done by someone in a subsequent PR

markov00 commented:
> Brush select also doesn't report events in Storybook (events get reported on the time charts though)

Fixed on c18ba7e
I've also fixed the brush extents on bar charts

monfera commented on src/chart_types/xy_chart/state/selectors/compute_brush_extent.ts:
```diff
@@ -0,0 +1,13 @@
+import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_specs';
+import { computeChartTransformSelector } from './compute_chart_transform';
+
+export const computeBrushExtentSelector = createCachedSelector(
```
Maybe I'm looking at it the wrong way, but don't immediately see where `computeBrushExtentSelector` is used

monfera commented on src/chart_types/xy_chart/state/selectors/compute_brush_extent.ts:
```diff
@@ -0,0 +1,13 @@
+import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_specs';
+import { computeChartTransformSelector } from './compute_chart_transform';
+
+export const computeBrushExtentSelector = createCachedSelector(
```
Maybe I'm looking at it the wrong way, but don't immediately see where `computeBrushExtentSelector` is used. Also, some other selectors eg. `isBrushingEnabledSelector`, `isCrosshairVisibleSelector`, `isLegendVisibleSelector`, `isTooltipEnabledSelector` and some non-selectors seem to be unused, eg.  `createOnChartRenderedCalled`, `buildArcRenderProps`, `isSettingsSpec` and tens of others. Based on their names and them being nested deep in code structures, they don't look like part of an API, nor are there tests. There's also a possibility that some unused exports might not show up as unused because they're covered with a test case

markov00 replied:
There can be for sure some unused selector here and there, I will check and remove them

markov00 replied:
I've cleaned the code a bit, removing most of the unused functions and some exports in e585909

monfera replied:
There's still about a dozen; resolution is that someone (@nickofthyme?) would remove or comment out unused functions

monfera commented on package.json:
```diff
@@ -153,10 +155,14 @@
     "mobx-react": "^5.4.3",
     "newtype-ts": "^0.2.4",
     "prop-types": "^15.7.2",
+    "re-reselect": "^3.4.0",
```
`reselect` caches the value for the last call only (by default, based on a shallow equality check).  

I haven't immediately seen a default cache size limitation in the `re-reselect` doc, is there something in the doc about a default limit? Question 3 [in the FAQ])https://github.com/toomuchdesign/re-reselect#faq) says that if we want to limit the number of cache entries, a `cacheObject` is needed, eg. `LruObjectCache`, but there don't appear to be any reference to these. It could mean that there's a memory leak, with memory growing over use. I haven't confirmed it and depending on selector update types and frequencies, it may or may not matter, but setting a limit may be good insurance

markov00 replied:
Good catch, it's using `FlatObjectCache` that can be "unlimited" depending on key.
The key used is the `chartId` that can be passed on the component or generated on every new chart instantiation. Using the `LruObjectCache` can be the right approach. What could be the cacheSize? we need to think to keep at least the number of charts in a dashboard so something around 10 to 50 visualizations?

monfera replied:
Resolved as @markov00 making a subsequent issue or PR for eg. a LRU cache with a limit; also, my initial concern was mitigated as I thought the proliferation of cached items will be pretty bad, but Marco explained about a key `re-reselect` point

monfera commented on package.json:
```diff
@@ -153,10 +155,14 @@
     "mobx-react": "^5.4.3",
     "newtype-ts": "^0.2.4",
     "prop-types": "^15.7.2",
+    "re-reselect": "^3.4.0",
```
One use for the the earlier suggestion of abstracting around the selectors with eg. `select`- instead of importing it directly from `re-reselect` in a bunch of files -  is that we could try what `re-reselect` buys us right now. For example, does performance deteriorate if we switch back to `reselect`.  

- are we utilizing the cache, given that the [doc](https://github.com/toomuchdesign/re-reselect#re-reselect-solution) says "[`cacheKey` is] by default a string or number but it can be anything depending on the chosen cache strategy (see cache objects docs)" while our selectors tend to use a good number of objects?
- if we are utilizing the cache, are we realizing the benefits of `re-reselect`, ie. does it lead to noticeably or measurably better performance?

On the offchance we're not utilizing the cache or it doesn't bring us benefits, we might want to temporarily revert to basic `reselect` until further measures. Should be a change of a few lines after abstracting out the selector.

One concern about abstracting out the selector could be that in the future, if/when we need it, we might want to use various caching strategies, ie. a variety of selectors could be needed. And it's a possible need in the future, eg. use a different strategy when the selector can be cached by primitive key vs. shallow object vs. deep object (eg. with expensive computations, it may be worth doing a deep equality check on a small object). But nothing stops us from having one `select` function that takes an optional second argument as the selector strategy, eg. 
```
import { select } from `ourSelectorAbstraction`;
...
const hoveredDataPoint = select(xyToDataPoint, {
      selector: `re-reselect`, 
      keys: `primitives`, 
      strategy: `LRU`, 
      limit: 100
  })(x, y);
```
or, to cut down on noise, via abstracting out common strategies, and using pre-configured shorthand literals, eg.
```
const hoveredDataPoint = select(xyToDataPoint, primitiveLruCache)(x, y);
```

monfera replied:
Again, subsequent PR by any one of us

markov00 commented:
> Render change action is unreported as well in Storybook

fixed on 0d93c57

markov00 commented:
> Crosshairs events don't get reported in either [here](https://elastic.github.io/elastic-charts/?path=/story/interactions--crosshair-with-time-axis) or in this branch, so maybe it's not meant to or regressed earlier

We don't have events from the crosshair itself, but there are events for what we call cursor position (https://elastic.github.io/elastic-charts/?path=/story/interactions--cursor-update-action)

rshen91 commented on src/chart_types/xy_chart/renderer/canvas/axis.tsx:
```diff
@@ -227,4 +214,108 @@ export class Axis extends React.PureComponent<AxisProps> {
+  }
+}
+
+const mapDispatchToProps = () => ({});
```
This is more of a question than stating that this needs to be fixed as it doesn't seem linked to any major performance issues. I was seeing in a couple files, `mapDispatchToProps` that don't seem to be used/needed. One example is here in axis, and the others I found are in the following: bar_values.tsx, grid.tsx, annotation_tooltip.tsx, brush.tsx, crosshair.tsx, highlighter.tsx, tooltip.tsx, and chart_status.tsx. Do these need to be included in the `connect()` at all? Or is it included for a reason? Like the redux structure to be in place incase we would want to build on it later? https://stackoverflow.com/questions/41670146/how-does-connect-work-without-mapdispatchtoprops

markov00 replied:
fix in ae398b5

nickofthyme commented on src/components/chart.tsx:
```diff
@@ -30,70 +38,70 @@ interface ChartProps {
+
+    const storeReducer = chartStoreReducer(uuid.v4());
+    if (process.env.NODE_ENV !== 'production') {
+      this.chartStore = createStore(storeReducer, devToolsEnhancer({ trace: true }));
```
👍

nickofthyme commented on src/components/chart.tsx:
```diff
@@ -30,70 +38,70 @@ interface ChartProps {
+    const onElementOverCaller = createOnElementOverCaller();
+    const onElementOutCaller = createOnElementOutCaller();
+    const onBrushEndCaller = createOnBrushEndCaller();
+    const onPointerMoveCaller = createOnPointerMoveCaller();
```
I think this is the best way to do this

nickofthyme commented on src/components/chart.tsx:
```diff
@@ -150,34 +158,21 @@ export class Chart extends React.Component<ChartProps, ChartState> {
-            <Tooltips getChartContainerRef={this.getChartContainerRef} />
-            <AnnotationTooltip getChartContainerRef={this.getChartContainerRef} />
-            <Highlighter />
+            <ChartContainer getChartContainerRef={this.getChartContainerRef} forwardStageRef={this.chartStageRef} />
```
Nice!

nickofthyme commented on src/components/chart_status.tsx:
```diff
@@ -0,0 +1,39 @@
+    const { onRenderChange, rendered } = this.props;
+    if (onRenderChange) {
+      onRenderChange(rendered);
+    }
```
Could be helper function on component

markov00 replied:
fixed in 4c9c7a04

nickofthyme commented on src/components/icons/assets/dot.tsx:
```diff
@@ -1,15 +1,17 @@
-      </g>
-    </svg>
-  );
+export class DotIcon extends React.PureComponent<Props> {
```
Is this for a reason? You could just use `memo` with the FC

markov00 replied:
Yes, but I've found that too late. I will take this in a future PR

nickofthyme replied:
No worries!!

nickofthyme commented on src/components/icons/icon.tsx:
```diff
@@ -31,25 +31,27 @@ export interface IconProps {
-export const Icon = (props: Props) => {
-  const { type, color, className, tabIndex, ...rest } = props;
-  let optionalCustomStyles = null;
+export class Icon extends React.PureComponent<Props> {
```
Is this for a reason? You could just use `memo` with the FC

markov00 replied:
Yes, but I've found that too late. I will take this in a future PR

nickofthyme commented on src/specs/settings.tsx:
```diff
@@ -99,135 +93,48 @@ export interface SettingSpecProps {
-  if (onRenderChange) {
-    chartStore.setOnRenderChangeListener(onRenderChange);
-  }
-}
```
Nice!!

nickofthyme commented on src/specs/specs_parser.tsx:
```diff
@@ -1,26 +1,37 @@
-export interface SpecProps {
-  chartStore?: ChartStore; // FIX
-}
+export const SpecsParserComponent: React.FunctionComponent<{}> = (props) => {
```
I think the type here is not necessary 

```suggestion
export const SpecsParserComponent = (props: DispatchProps) => {
```

markov00 replied:
It's required to type the `children` property of props

nickofthyme commented on src/state/selectors/get_deselected_data_series.ts:
```diff
@@ -0,0 +1,16 @@
+// import createCachedSelector from 're-reselect';
```
?

markov00 replied:
removed in 7d74a5a

nickofthyme commented on src/state/selectors/get_legend_size.ts:
```diff
@@ -0,0 +1,96 @@
+const VALUE_LEFT_MARGIN = 4;
+const VERTICAL_PADDING = 4;
+
+export const getLegendSizeSelector = createCachedSelector(
```
Nice!

nickofthyme commented on src/state/spec_factory.ts:
```diff
@@ -0,0 +1,61 @@
+    useEffect(() => {
+      // console.log('updating specs');
+      if (prevId && prevId !== props.id) {
+        // console.log('remove existing specs, ', prevId);
```
```suggestion
```

markov00 replied:
fixed in 76e7d34

nickofthyme commented on src/state/spec_factory.ts:
```diff
@@ -0,0 +1,61 @@
+    const prevId = usePrevious(props.id);
+    const { removeSpec, upsertSpec, ...spec } = props;
+    useEffect(() => {
+      // console.log('updating specs');
```
```suggestion
```

markov00 replied:
fixed in 76e7d34

nickofthyme commented on src/state/spec_factory.ts:
```diff
@@ -0,0 +1,61 @@
+    });
+    useEffect(
+      () => () => {
+        // console.log('remove spec on unmount', props.id);
```
```suggestion
```

markov00 replied:
fixed in 76e7d34

nickofthyme commented on src/utils/bbox/dom_text_bbox_calculator.ts:
```diff
@@ -0,0 +1,39 @@
+import { Option, some } from 'fp-ts/lib/Option';
```
What is `fp-ts` doing?

markov00 replied:
This is a nice functional programming library that provides some nice types. In this case returning an Option types mean that, on the result you can call `getOrElse` and specify a default instead of checking for null values or similar.
I've removed that here 14d7a946 because right now it's not really useful. I've moved the default return value (used by all the callers) inside the calculator function.

nickofthyme replied:
oh I see. Reminds me of the little scala I used.

nickofthyme commented on src/utils/ids.ts:
```diff
@@ -1,25 +1,17 @@
 }
-export function getAxisId(id: string): AxisId {
-  return axisIdIso.wrap(id);
+export function getAxisId(id: string): string {
```
Feels so good!

markov00 commented on integration/page_objects/common.ts:
```diff
@@ -21,7 +21,7 @@ class CommonPage {
     return `${baseUrl}?${query}${query ? '&' : ''}knob-debug=false`;
   }
-  async getBoundingClientRect(selector = '.echChart[data-ech-render-complete=true]') {
+  async getBoundingClientRect(selector = '.echChart') {
```
No this is because I've moved the `data render` attributes into a child element as written on the breaking changes notes in the PR description. The reason was to separate the two element to avoid unnecessary re-rendering

markov00 commented on src/chart_types/index.ts:
```diff
@@ -0,0 +1,5 @@
+  Global: 'global' as 'global',
+  Pie: 'pie' as 'pie',
+  XYAxis: 'xy_axis' as 'xy_axis',
+});
```
fixed in 5fc93cff

markov00 commented on src/chart_types/xy_chart/legend/legend.ts:
```diff
@@ -40,20 +40,31 @@ export function getPostfix(spec: BasicSeriesSpec): Postfixes {
   sortedSeriesColors.forEach((series, key) => {
     const { banded, specId, lastValue, colorValues } = series;
-    const spec = specs.get(specId);
+    const spec = getSpecsById<BasicSeriesSpec>(specs, specId);
```
I'm not much a fan of classes, but we can create for sure a set of utilities for that!

markov00 commented on src/chart_types/xy_chart/renderer/canvas/axis.tsx:
```diff
@@ -9,13 +9,24 @@ import {
+import { computeAxisTicksDimensionsSelector } from '../../state/selectors/compute_axis_ticks_dimensions';
+import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
+import { computeChartDimensionsSelector } from '../../state/selectors/compute_chart_dimensions';
+import { LIGHT_THEME } from '../../../../utils/themes/light_theme';
```
Absolutely, I've added aliases at the beginning but it creates a bit of issues sometimes when moving a lot of files and I always have to restart TS Server on vscode for that

markov00 commented on src/chart_types/xy_chart/renderer/canvas/reactive_chart.tsx:
```diff
@@ -0,0 +1,390 @@
+    ];
+  };
+
+  // getAxes = (): AxisProps[] => {
```
fix in a2540567

markov00 commented on src/chart_types/xy_chart/renderer/canvas/index.tsx:
```diff
@@ -0,0 +1,25 @@
+// renderBrushTool = () => {
```
removed in a2540567

markov00 commented on src/chart_types/xy_chart/renderer/canvas/reactive_chart.tsx:
```diff
@@ -0,0 +1,390 @@
+}
+interface ReactiveChartElementIndex {
+  element: JSX.Element;
+  zIndex: number;
```
It's mostly used to change the zIndex of the annotation in respect of the geometries. We can also apply it to the series if we want

markov00 commented on src/chart_types/xy_chart/renderer/canvas/line_annotation.tsx:
```diff
@@ -1,19 +1,18 @@
+
+    return <Group key={'line_annotations'}>{lines.map(this.renderAnnotationLine)}</Group>;
   }
   private renderAnnotationLine = (lineConfig: AnnotationLineProps, i: number) => {
```
renamed in a697b928

markov00 commented on src/chart_types/xy_chart/renderer/canvas/line_annotation.tsx:
```diff
@@ -1,19 +1,18 @@
-    return this.renderAnnotation();
+    const { lines } = this.props;
+
+    return <Group key={'line_annotations'}>{lines.map(this.renderAnnotationLine)}</Group>;
```
removed in a697b928 (together with the same in rect_annotations)

markov00 commented on src/chart_types/xy_chart/renderer/canvas/reactive_chart.tsx:
```diff
@@ -0,0 +1,390 @@
+
+interface ReactiveChartStateProps {
+  initialized: boolean;
+  geometries: {
```
fix on 95276b69

markov00 commented on src/chart_types/xy_chart/renderer/canvas/rect_annotation.tsx:
```diff
@@ -0,0 +1,29 @@
+      height,
+    };
+
+    return <Rect {...rectProps} key={`rect-annotation-${i}`} />;
```
what do you mean by directly? on props at time like `x={x} y={y} ` etc?

nickofthyme replied:
Yeah exactly. I just feel like defining the props that you then spread into the component is silly, when you can just set the props on the component itself. Then you will see what props are missing directly.

No need to fix now.

markov00 commented on src/chart_types/xy_chart/renderer/dom/tooltips.tsx:
```diff
@@ -67,22 +81,21 @@ class TooltipsComponent extends React.Component<TooltipProps> {
       tooltip = (
         <div className="echTooltip" ref={this.tooltipRef}>
-          <div className="echTooltip__header">{this.renderHeader(tooltipData[0], tooltipHeaderFormatter)}</div>
+          <div className="echTooltip__header">{this.renderHeader(tooltipValues[0], tooltipHeaderFormatter)}</div>
```
I've opened an issue to track this: https://github.com/elastic/elastic-charts/issues/460

markov00 commented on src/chart_types/xy_chart/rendering/rendering.areas.test.ts:
```diff
@@ -1,19 +1,21 @@
-/* eslint @typescript-eslint/no-object-literal-type-assertion: off */
-import { computeSeriesDomains } from '../store/utils';
-import { getGroupId, getSpecId, SpecId } from '../../../utils/ids';
+/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
```
fix in 555db69c I've removed the eslint disable line without needs to use forceTypes in tests

nickofthyme replied:
Yeah I think it wasn't the exact case here that I was referring to.

markov00 commented on src/chart_types/xy_chart/state/chart_state.tsx:
```diff
@@ -0,0 +1,49 @@
+        <BrushTool />
+      </React.Fragment>
+    );
+  }
```
your comment is much more like black magic ?

markov00 commented on src/chart_types/xy_chart/specs/rect_annotation.tsx:
```diff
@@ -1,33 +1,24 @@
-type RectAnnotationProps = SpecProps & RectAnnotationSpec;
+const defaultProps = {
+  chartType: ChartTypes.XYAxis,
+  specType: 'annotation' as 'annotation',
```
fix on 56992787

markov00 commented on src/chart_types/xy_chart/specs/rect_annotation.tsx:
```diff
@@ -1,33 +1,24 @@
+  chartType: ChartTypes.XYAxis,
+  specType: 'annotation' as 'annotation',
+  groupId: DEFAULT_GLOBAL_ID,
+  annotationType: 'rectangle' as 'rectangle',
```
fix on 56992787

nickofthyme replied:
Love it!! : 😍

markov00 commented on src/chart_types/xy_chart/specs/histogram_bar_series.tsx:
```diff
@@ -1,42 +1,36 @@
+const defaultProps = {
+  chartType: ChartTypes.XYAxis,
+  specType: 'series' as 'series',
+  seriesType: 'bar' as 'bar',
```
fixed on 56992787

markov00 commented on src/chart_types/xy_chart/state/chart_state.interactions.test.ts:
```diff
@@ -0,0 +1,657 @@
+    expect(seriesGeoms.scales.yScales).not.toBeUndefined();
+  });
+
+  // test('set cursor from external source', () => {
```
fixed on 018883b9
I've skipped them but commented the inside code because it's related to the old implementation. I will take that in a future PR

nickofthyme replied:
Oh I see, where there would be TS issues? Got it.

markov00 commented on src/chart_types/xy_chart/state/utils.ts:
```diff
@@ -72,6 +64,34 @@ export interface GeometriesCounts {
+
+export interface ComputedGeometries {
+  scales: ComputedScales;
+  geometries: {
```
fixed in 95276b696406b262fd6872ceffe0e4fd7a8add09

markov00 commented on src/chart_types/xy_chart/state/selectors/merge_y_custom_domains.ts:
```diff
@@ -0,0 +1,71 @@
+      domainsByGroupId.set(groupId, domain);
+    }
+  });
+  // console.log('--- 0 mergeYCustomDomains ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/is_chart_animatable.ts:
```diff
@@ -0,0 +1,17 @@
+export const isChartAnimatableSelector = createCachedSelector(
+  [computeSeriesGeometriesSelector, getSettingsSpecSelector],
+  () => {
+    // console.log('--- 13 isChartAnimatableSelector ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/get_series_color_map.ts:
```diff
@@ -0,0 +1,22 @@
+      chartTheme.colors,
+      updatedCustomSeriesColors,
+    );
+    // console.log('--- 2 computeSeriesDomainsSelector ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/count_bars_in_cluster.ts:
```diff
@@ -0,0 +1,15 @@
+    const { formattedDataSeries } = seriesDomainsAndData;
+
+    const { totalBarsInCluster } = countBarsInCluster(formattedDataSeries.stacked, formattedDataSeries.nonStacked);
+    // console.log('--- 4 countBarsInCluster ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_series_geometries.ts:
```diff
@@ -0,0 +1,48 @@
+    axesSpecs,
+    isHistogramMode,
+  ): ComputedGeometries => {
+    // console.log('--- 9 computeSeriesGeometriesSelector ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_series_domains.ts:
```diff
@@ -0,0 +1,24 @@
+export const computeSeriesDomainsSelector = createCachedSelector(
+  [getSeriesSpecsSelector, mergeYCustomDomainsByGroupIdSelector, getDeselectedSeriesSelector, getSettingsSpecSelector],
+  (seriesSpecs, customYDomainsByGroupId, deselectedDataSeries, settingsSpec): SeriesDomainsAndData => {
+    // console.log('--- 1 computeSeriesDomainsSelector ---', seriesSpecs);
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_legend.ts:
```diff
@@ -0,0 +1,40 @@
+    axesSpecs,
+    deselectedDataSeries,
+  ): Map<string, LegendItem> => {
+    // console.log('--- 3 computeLegend ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_chart_transform.ts:
```diff
@@ -0,0 +1,13 @@
+export const computeChartTransformSelector = createCachedSelector(
+  [computeChartDimensionsSelector, getSettingsSpecSelector],
+  (chartDimensions, settingsSpecs): Transform => {
+    // console.log('--- 7 computeChartTransformSelector ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_axis_visible_ticks.ts:
```diff
@@ -0,0 +1,60 @@
+    isHistogramMode,
+    barsPadding,
+  ): AxisVisibleTicks => {
+    // console.log('--- 10 computeAxisVisibleTicksSelector ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_axis_ticks_dimensions.ts:
```diff
@@ -0,0 +1,61 @@
+      }
+    });
+    bboxCalculator.destroy();
+    // console.log('--- 5 computeAxisTickDimensions ---');
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/state/selectors/compute_annotations.ts:
```diff
@@ -0,0 +1,45 @@
+    totalBarsInCluster,
+    isHistogramMode,
+  ): Map<AnnotationId, AnnotationDimensions> => {
+    // console.log('--- 11 computeAnnotationDimensionsSelector ---', annotationSpecs);
```
fixed in 76e7d346

markov00 commented on src/chart_types/xy_chart/annotations/line_annotation_tooltip.ts:
```diff
@@ -0,0 +1,438 @@
+  if (isHorizontalAxis(axisPosition)) {
+    xOffset = isChartHorizontalRotation ? 50 : 0;
+  } else {
+    xOffset = isChartHorizontalRotation ? (axisPosition === Position.Right ? 100 : 0) : 50;
```
fixed in 9b0bb9fa

markov00 commented on src/chart_types/xy_chart/annotations/rect_annotation_tooltip.ts:
```diff
@@ -0,0 +1,169 @@
+
+export function isWithinRectBounds(
+  cursorPosition: Point,
+  { startX, endX, startY, endY }: { startX: number; endX: number; startY: number; endY: number },
```
fixed in 9b0bb9fa

markov00 commented on src/chart_types/xy_chart/annotations/line_annotation_tooltip.ts:
```diff
@@ -0,0 +1,438 @@
+export function isWithinLineMarkerBounds(cursorPosition: Point, marker: AnnotationMarker): boolean {
+  const { top, left } = marker.position;
+  const { width, height } = marker.dimension;
+  const markerRect = { startX: left, startY: top, endX: left + width, endY: top + height };
```
fixed in 9b0bb9fa

markov00 commented on src/chart_types/xy_chart/annotations/rect_annotation_tooltip.ts:
```diff
@@ -0,0 +1,169 @@
+}
+
+export function isWithinRectBounds(
+  cursorPosition: Point,
```
fixed in 9b0bb9fa

markov00 commented on src/chart_types/xy_chart/renderer/canvas/rect_annotation.tsx:
```diff
@@ -0,0 +1,29 @@
+    const { rects } = this.props;
+    return <Group key={'rect_annotations'}>{rects.map(this.renderAnnotationRect)}</Group>;
+  }
+  private renderAnnotationRect = (props: AnnotationRectProps, i: number) => {
```
fix in 428151ae

markov00 commented:
@nickofthyme I've fixed the Annotation tooltips here: 143276cc
I've left the `shift+click` for a future PR

nickofthyme commented:
YAAAAY!!!!

markov00 commented:
:tada: This PR is included in version 15.0.0 :tada:

The release is available on:
- [GitHub release](https://github.com/elastic/elastic-charts/releases/tag/v15.0.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@elastic/charts)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
